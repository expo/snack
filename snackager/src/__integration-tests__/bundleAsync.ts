import fs from 'fs';
import glob from 'glob';
import mapValues from 'lodash/mapValues';
import os from 'os';
import path from 'path';
import util from 'util';

import { getCoreExternals } from '../bundler/externals';
import fetchAndExtract from '../utils/fetchAndExtract';
import fetchMetadata from '../utils/fetchMetadata';
import findVersion from '../utils/findVersion';
import getBundleInfo, { BundleInfo } from '../utils/getBundleInfo';
import installPackage from '../utils/installPackage';
import packageBundle from '../utils/packageBundle';
import parseRequest from '../utils/parseRequest';
import resolveDependencies from '../utils/resolveDependencies';

// TODO: replace rimraf with fs.rm once node 14.40 lands
const rimraf = util.promisify(require('rimraf'));

const DEFAULT_PLATFORMS = ['ios', 'android', 'web'];

type BundledPackage = {
  name: string;
  version: string;
  // TODO: fix possible null dependency and replace with `Package['peerDependencies']`
  peerDependencies: { [dependency: string]: string | null };
  files: { [platform: string]: { [file: string]: BundleInfo } };
};

// Keep track of all tested packages, to clean the lockfiles if they aren't tested anymore
const testedPackages: string[] = [];

export default async function bundleAsync(
  packageSpec: string,
  bundlePlatforms: string[] = DEFAULT_PLATFORMS,
  includeCode?: boolean
): Promise<BundledPackage> {
  const { qualified, id, tag, scope, deep, platforms } = parseRequest(
    `/${packageSpec}?platforms=${bundlePlatforms.join(',')}`
  );
  const workdir = path.join(
    fs.realpathSync(os.tmpdir()),
    'snackager',
    '__integration-tests__',
    qualified
  );
  try {
    await rimraf(workdir);
  } catch (e) {}
  const packagedir = path.join(workdir, 'package');

  const meta = await fetchMetadata(qualified, { scope, id });
  const { version, isLatest } = findVersion(qualified, meta, tag);
  const { pkg, dependencies } = resolveDependencies(meta, version, isLatest, deep);

  await fetchAndExtract(pkg, version, workdir);

  // keep track of the used lockfiles, and try to restore the lockfile
  // this represents the package + nested dependency "state" that was used in the bundle snapshot
  testedPackages.push(packageSpec);
  const cacheHit = await restoreLockfile(packageSpec, packagedir);
  if (cacheHit) {
    await installPackage(packagedir, ['--frozen-lockfile']);
  } else {
    await installPackage(packagedir);
    await saveLockfile(packageSpec, packagedir);
  }

  const files = await packageBundle({
    pkg,
    cwd: packagedir,
    platforms,
    base: 'nonsense',
    deep,
    externalDependencies: dependencies,
  });

  return {
    name: qualified,
    version,
    files: mapValues(files, (file, platform) =>
      mapValues(file, (buffer, filename) => {
        const metadata = getBundleInfo(filename, buffer, includeCode);
        (metadata.externals ?? []).forEach((external) => {
          if (!getCoreExternals().includes(external) && !dependencies[external]) {
            console.warn(
              `Bundle "${qualified}@${version}/${platform}-${filename}" contains external "${external}" which is not listed as a peer dependency in package.json`
            );
          }
        });
        return metadata;
      })
    ),
    peerDependencies: dependencies,
  };
}

// Location of the lockfiles to store them in, they should be considered similar to snapshots
const lockfileStorage = path.join(__dirname, '__snapshots__', 'lockfiles');

async function restoreLockfile(packageSpec: string, directory: string): Promise<boolean> {
  const packagePath = path.join(directory, 'yarn.lock');
  const storagePath = path.join(lockfileStorage, `${packageSpec}.lock`);

  try {
    await fs.promises.copyFile(storagePath, packagePath);
    // Lockfile is restored successfully (cache hit)
    return true;
  } catch {
    // Lockfile couldn't be restored (no cache hit)
    return false;
  }
}

async function saveLockfile(packageSpec: string, directory: string): Promise<void> {
  const packagePath = path.join(directory, 'yarn.lock');
  const storagePath = path.join(lockfileStorage, `${packageSpec}.lock`);

  await fs.promises.mkdir(path.dirname(storagePath), { recursive: true });
  // Always keep the lockfile up to date after an install
  await fs.promises.copyFile(packagePath, storagePath);

  console.warn(`Added lockfile for ${packageSpec}, commit this to the repository.`);
}

export async function cleanUnusedLockfiles(): Promise<void> {
  const existingPackages = await new Promise<string[]>((resolve, reject) => {
    glob('**/*.lock', { cwd: lockfileStorage }, (error, files) => {
      if (error) {
        reject(error);
      } else {
        resolve(files.map((filename) => filename.replace('.lock', '')));
      }
    });
  });

  const unusedPackages = existingPackages.filter(
    (packageSpec) => !testedPackages.includes(packageSpec)
  );

  // remove unused package lockfiles from the repo
  for (const packageSpec of unusedPackages) {
    await fs.promises.unlink(path.join(lockfileStorage, `${packageSpec}.lock`));
    console.warn(`Removed obsolete lockfile for ${packageSpec}, commit this to the repository.`);
  }
}
