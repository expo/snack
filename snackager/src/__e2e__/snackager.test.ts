import { bundleAsync, normalizeBundleSize } from './bundleAsync';

it('creates bundle for a single platform', async () => {
  const bundle = await bundleAsync('firestorter@2.0.1', ['ios']);
  expect(normalizeBundleSize(bundle)).toMatchSnapshot();
});

it('fails when no package name is specified', async () => {
  await expect(bundleAsync('', ['ios'])).rejects.toEqual(new Error(`Failed to parse request`));
});

it('externalizes dependencies that are marked as external', async () => {
  // `expo-google-app-auth` declares `expo-app-auth` as a direct dependency.
  // The `expo-app-auth` dependency should however still be externalized and
  // not linked into the bundle.
  const bundle = await bundleAsync('expo-google-app-auth@8.1.3');
  expect(normalizeBundleSize(bundle)).toMatchSnapshot();
  expect(bundle.files.android['bundle.js'].externals).toEqual(
    expect.arrayContaining(['expo-app-auth']),
  );
});

it('resolves external source-code references', async () => {
  // `expo-linear-gradient` imports source-code from react-native-web directly.
  // e.g. "import normalizeColor from 'react-native-web/src/modules/normalizeColor';"
  // The bundler should install `react-native-web` and resolve the import correctly.
  // All regular `react-native-web` imports should be externalized.
  const bundle = await bundleAsync('expo-linear-gradient@8.2.1');
  expect(normalizeBundleSize(bundle)).toMatchSnapshot();
});

it('resolves commonly missing peer-deps (prop-types)', async () => {
  // For example 'react-native-responsive-grid@0.32.4' has a dependency
  // on prop-types but no peerDep listed in package.json.
  const bundle = await bundleAsync('react-native-responsive-grid@0.32.4');
  expect(normalizeBundleSize(bundle)).toMatchSnapshot();
  expect(bundle.files.android['bundle.js'].externals).not.toEqual(
    expect.arrayContaining(['prop-types']),
  );
});

it('disallows bundling of core modules', async () => {
  await expect(bundleAsync('expo', ['ios'])).rejects.toEqual(
    new Error(`Bundling core module 'expo' is prohibited`),
  );
  await expect(bundleAsync('react-native', ['ios'])).rejects.toEqual(
    new Error(`Bundling core module 'react-native' is prohibited`),
  );
  await expect(bundleAsync('react-native-web', ['ios'])).rejects.toEqual(
    new Error(`Bundling core module 'react-native-web' is prohibited`),
  );
  await expect(bundleAsync('react-native-windows', ['ios'])).rejects.toEqual(
    new Error(`Bundling core module 'react-native-windows' is prohibited`),
  );
  await expect(bundleAsync('react-native/Libraries/Image/AssetRegistry', ['ios'])).rejects.toEqual(
    new Error(`Bundling core module 'react-native/Libraries/Image/AssetRegistry' is prohibited`),
  );
});
