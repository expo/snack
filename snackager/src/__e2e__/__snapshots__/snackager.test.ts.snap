// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates bundle for a single platform 1`] = `
{
  "files": {
    "ios": {
      "bundle.js": {
        "externals": [
          "mobx",
        ],
        "size": 42000,
      },
    },
  },
  "name": "firestorter",
  "peerDependencies": {
    "mobx": "^4.0.0",
  },
  "version": "2.0.1",
}
`;

exports[`externalizes dependencies that are marked as external 1`] = `
{
  "files": {
    "android": {
      "bundle.js": {
        "externals": [
          "@unimodules/core",
          "expo-app-auth",
          "expo-constants",
          "react-native",
        ],
        "size": 6000,
      },
    },
    "ios": {
      "bundle.js": {
        "externals": [
          "@unimodules/core",
          "expo-app-auth",
          "expo-constants",
          "react-native",
        ],
        "size": 6000,
      },
    },
    "web": {
      "bundle.js": {
        "externals": [
          "@unimodules/core",
          "expo-app-auth",
          "expo-constants",
          "react-native",
        ],
        "size": 6000,
      },
    },
  },
  "name": "expo-google-app-auth",
  "peerDependencies": {
    "@unimodules/core": "*",
    "expo-app-auth": "~9.2.0",
    "expo-constants": "*",
    "react": "*",
    "react-native": "*",
  },
  "version": "8.1.3",
}
`;

exports[`resolves commonly missing peer-deps (prop-types) 1`] = `
{
  "files": {
    "android": {
      "bundle.js": {
        "externals": [
          "react",
          "react-native",
          "react/jsx-runtime",
        ],
        "size": 18000,
      },
    },
    "ios": {
      "bundle.js": {
        "externals": [
          "react",
          "react-native",
          "react/jsx-runtime",
        ],
        "size": 18000,
      },
    },
    "web": {
      "bundle.js": {
        "externals": [
          "react",
          "react-native",
          "react/jsx-runtime",
        ],
        "size": 18000,
      },
    },
  },
  "name": "react-native-responsive-grid",
  "peerDependencies": {},
  "version": "0.32.4",
}
`;

exports[`resolves external source-code references 1`] = `
{
  "files": {
    "android": {
      "bundle.js": {
        "externals": [
          "@unimodules/core",
          "react",
          "react-native",
        ],
        "size": 6000,
      },
    },
    "ios": {
      "bundle.js": {
        "externals": [
          "@unimodules/core",
          "react",
          "react-native",
        ],
        "size": 6000,
      },
    },
    "web": {
      "bundle.js": {
        "externals": [
          "react",
          "react-native",
        ],
        "size": 15000,
      },
    },
  },
  "name": "expo-linear-gradient",
  "peerDependencies": {
    "react": "*",
    "react-native": "*",
  },
  "version": "8.2.1",
}
`;
